libs/wire-api/src/Wire/API/Routes/Public/Brig/OAuth.hs

```diff
@@ -0,0 +1,160 @@
+           )
+    :<|> Named
+           "get-oauth-applications"
+           ( Summary "Get OAuth applications with account access"
+               :> Description "Get all OAuth applications with active account access for a user."
+               :> ZUser
+               :> "oauth"
+               :> "applications"
+               :> MultiVerb1
+                    'GET
+                    '[JSON]
+                    (Respond 200 "OAuth applications found" [OAuthApplication])
+           )
+    :<|> Named
+           "revoke-oauth-account-access"
+           ( Summary "Revoke account access from an OAuth application"
+               :> ZUser
+               :> "oauth"
+               :> "applications"
+               :> Capture' '[Description "The ID of the OAuth client"] "OAuthClientId" OAuthClientId
+               :> MultiVerb
+                    'DELETE
+                    '[JSON]
+                    '[RespondEmpty 204 "OAuth application access revoked"]
+                    ()
+           )
+
+type CreateOAuthAuthCodeHeaders = '[Header "Location" RedirectUrl]
+
+type CreateOAuthAuthCodeResponses =
+  '[ -- success
+     WithHeaders CreateOAuthAuthCodeHeaders RedirectUrl (RespondEmpty 302 "Found"),
+     -- feature disabled
+     WithHeaders CreateOAuthAuthCodeHeaders RedirectUrl (RespondEmpty 403 "Forbidden"),
+     -- unsupported response type
+     WithHeaders CreateOAuthAuthCodeHeaders RedirectUrl (RespondEmpty 400 "Bad Request"),
+     -- client not found
+     WithHeaders CreateOAuthAuthCodeHeaders RedirectUrl (RespondEmpty 404 "Not Found"),
+     -- redirect url mismatch
+     ErrorResponse 'OAuthRedirectUrlMissMatch
+   ]
+
+data CreateOAuthCodeResponse
+  = CreateOAuthCodeSuccess RedirectUrl
+  | CreateOAuthCodeFeatureDisabled RedirectUrl
+  | CreateOAuthCodeUnsupportedResponseType RedirectUrl
+  | CreateOAuthCodeClientNotFound RedirectUrl
+  | CreateOAuthCodeRedirectUrlMissMatch
+
+instance AsUnion CreateOAuthAuthCodeResponses CreateOAuthCodeResponse where
+  toUnion :: CreateOAuthCodeResponse -> Union (ResponseTypes CreateOAuthAuthCodeResponses)
+  toUnion (CreateOAuthCodeSuccess url) = Z (I url)
+  toUnion (CreateOAuthCodeFeatureDisabled url) = S (Z (I url))
+  toUnion (CreateOAuthCodeUnsupportedResponseType url) = S (S (Z (I url)))
+  toUnion (CreateOAuthCodeClientNotFound url) = S (S (S (Z (I url))))
+  toUnion CreateOAuthCodeRedirectUrlMissMatch = S (S (S (S (Z (I (dynError @(MapError 'OAuthRedirectUrlMissMatch)))))))
+  fromUnion :: Union (ResponseTypes CreateOAuthAuthCodeResponses) -> CreateOAuthCodeResponse
+  fromUnion (Z (I url)) = CreateOAuthCodeSuccess url
+  fromUnion (S (Z (I url))) = CreateOAuthCodeFeatureDisabled url
+  fromUnion (S (S (Z (I url)))) = CreateOAuthCodeUnsupportedResponseType url
+  fromUnion (S (S (S (Z (I url))))) = CreateOAuthCodeClientNotFound url
+  fromUnion (S (S (S (S (Z (I _)))))) = CreateOAuthCodeRedirectUrlMissMatch
+  fromUnion (S (S (S (S (S x))))) = case x of {}
```

Comment:
what the?!  isn't there a less boilerplaty way of doing this?

Reply:
:shrug: